ffi/src/lib.rs

```diff
@@ -790,212 +790,6 @@ pub unsafe extern "C" fn free_string_slice_data(data: Handle<StringSliceIterator
-    fn visit_struct_fields(visitor: &EngineSchemaVisitor, s: &StructType) -> usize {
-        let child_list_id = (visitor.make_field_list)(visitor.data, s.fields.len());
-        for field in s.fields() {
-            visit_schema_item(field.data_type(), field.name(), visitor, child_list_id);
-        }
-        child_list_id
-    }
-
-    fn visit_array_item(visitor: &EngineSchemaVisitor, at: &ArrayType) -> usize {
-        let child_list_id = (visitor.make_field_list)(visitor.data, 1);
-        visit_schema_item(&at.element_type, "array_element", visitor, child_list_id);
-        child_list_id
-    }
-
-    fn visit_map_types(visitor: &EngineSchemaVisitor, mt: &MapType) -> usize {
-        let child_list_id = (visitor.make_field_list)(visitor.data, 2);
-        visit_schema_item(&mt.key_type, "map_key", visitor, child_list_id);
-        visit_schema_item(&mt.value_type, "map_value", visitor, child_list_id);
-        child_list_id
-    }
-
-    // Visit a struct field (recursively) and add the result to the list of siblings.
-    fn visit_schema_item(
-        data_type: &DataType,
-        name: &str,
-        visitor: &EngineSchemaVisitor,
-        sibling_list_id: usize,
-    ) {
-        macro_rules! call {
-            ( $visitor_fn:ident $(, $extra_args:expr) *) => {
-                (visitor.$visitor_fn)(visitor.data, sibling_list_id, name.into() $(, $extra_args) *)
-            };
-        }
-        match data_type {
-            DataType::Struct(st) => call!(visit_struct, visit_struct_fields(visitor, st)),
-            DataType::Map(mt) => {
-                call!(
-                    visit_map,
-                    mt.value_contains_null,
-                    visit_map_types(visitor, mt)
-                )
-            }
-            DataType::Array(at) => {
-                call!(visit_array, at.contains_null, visit_array_item(visitor, at))
-            }
-            DataType::Primitive(PrimitiveType::Decimal(precision, scale)) => {
-                call!(visit_decimal, *precision, *scale)
-            }
-            &DataType::STRING => call!(visit_string),
-            &DataType::LONG => call!(visit_long),
-            &DataType::INTEGER => call!(visit_integer),
-            &DataType::SHORT => call!(visit_short),
-            &DataType::BYTE => call!(visit_byte),
-            &DataType::FLOAT => call!(visit_float),
-            &DataType::DOUBLE => call!(visit_double),
-            &DataType::BOOLEAN => call!(visit_boolean),
-            &DataType::BINARY => call!(visit_binary),
-            &DataType::DATE => call!(visit_date),
-            &DataType::TIMESTAMP => call!(visit_timestamp),
-            &DataType::TIMESTAMP_NTZ => call!(visit_timestamp_ntz),
-        }
-    }
-
-    visit_struct_fields(visitor, snapshot.schema())
-}
-
-// TODO move expression visitors to separate module
-
 // A set that can identify its contents by address
 pub struct ReferenceSet<T> {
```

Comment:
aside: Right now, this is only used by the expression visitors. 
Should we move it there? Or do we anticipate it being useful more generally?

Reply:
I think this would be useful in any Engine -> Kernel conversions, like how its used for expressions. Right now, I'm not sure if we'd have usecases where the engine must convert anything besides expressions. Perhaps schemas or some other metadata?